### PR TITLE
[TRTLLM-7155][feat] Unify sampler handle logits implementation.

### PR DESCRIPTION
**Pull request by dcampora**
## Summary by CodeRabbit

* Refactor
  * Unified beam width handling for consistent sampling across requests.
  * Streamlined logits processing; computed only when requested to reduce overhead.
  * Generation logits are no longer returned; outputs focus on token log-probabilities.
  * Removed host-side logits storage to lower memory usage and improve efficiency.
  * More consistent behavior when overlap scheduling is enabled.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tests/unittest/_torch/test_return_logits.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 global_kvcache_config = KvCacheConfig(max_tokens=10000)
 
 
-@force_ampere  # Save H100 resource
+# @force_ampere  # Save H100 resource
```
[the remark this answers is not in the document]
Comment (dcampora) [reply tests/unittest/_torch/test_return_logits.py:15]: No, thanks for catching that. That's for my local testing. I'll revert it.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 2]
```diff
@@ -832,18 +810,10 @@ def _executor_loop_pp(self):
                         torch.cuda.nvtx.range_push(
                             "_handle_new_tokens_inter_pp")
                         # Receive tokens from previous pp rank (w.r.t model forward direction)
-                        (
-                            logits,
-                            sample_state.host,
-                        ) = self.dist.recv_object(
+                        (sample_state.host, ) = self.dist.recv_object(
```
Comment (yuxianq) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:813]: Can send/recv `sample_state.host` instead of `(sample_state.host, )` directly.

Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py:813]: Done.

**Review by mikeiovine — APPROVED** [review]
All speculative decoding related code looks good to me - leaving rest of the review for others with more context

**Review by DomBrown — APPROVED** [review]

**Review by yuxianq — APPROVED** [review]
The pp+logits part LGTM.

File: tensorrt_llm/_torch/pyexecutor/handle_logits.py [hunk 3]
```diff
@@ -11,25 +11,37 @@ class HandleLogits:
...
-        generation_requests: List[LlmRequest],
-        logits: torch.Tensor,
-        num_context_logits_prefix_sum: List[int],
-        max_num_sequences: int,
-        beam_width: int,
-    ):
+    def __call__(self, context_requests: List[LlmRequest],
+                 generation_requests: List[LlmRequest], logits: torch.Tensor,
+                 beam_width: int, num_context_logits_prefix_sum: list[int],
+                 is_generation_model: bool):
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/handle_logits.py:17]: Add comma to keep formatting.

File: tensorrt_llm/_torch/pyexecutor/handle_logits.py [hunk 4]
```diff
@@ -11,25 +11,37 @@ class HandleLogits:
...
         """
+        if not any(r.py_return_context_logits
+                   for r in context_requests) and not any(
+                       r.py_return_generation_logits
+                       for r in generation_requests):
+            return
+
+        if not is_generation_model:
+            for batch_index, llm_req in enumerate(context_requests):
+                logits_temp = logits[batch_index]
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/handle_logits.py:36]:
```suggestion
            for llm_req, logits_temp in zip(context_requests, logits):
```

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 5]
```diff
@@ -884,6 +845,40 @@ def _executor_loop_pp(self):
...
+
+                        elif self.dist.is_last_pp_rank and len(finished_reqs):
+                            if self.send_handles[
+                                    prev_microbatch_id] is not None:
+                                self.send_handles[prev_microbatch_id].wait()
+                            self.send_handles[
+                                prev_microbatch_id] = self.dist.isend_object(
+                                    [r.py_result for r in finished_reqs],
+                                    dest=self.dist.next_pp_rank,
+                                    tag=prev_microbatch_id)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:880]: Can we move the communication logic to a separate function, please?

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 6]
```diff
@@ -659,19 +655,35 @@ def _apply_embedding_bias(
...
-        raw_logits = model_outputs["logits"]
+
+        # raw_logits should contain only the logits from the gen requests.
+        # If return context logits is requested, fetch only the logits from gen requests.
+        if any(r.py_return_context_logits
+               for r in scheduled_requests.context_requests):
+            gen_logits_indices = []
+            total_context_logits = num_context_logits_prefix_sum[-1]
+            for i in range(len(scheduled_requests.context_requests)):
+                gen_logits_indices.append(num_context_logits_prefix_sum[i + 1] -
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:674]: If `num_context_logits_prefix_sum` was an numpy array we could simply write
```
gen_logits_indices.extend(num_context_logits_prefix_sum[1:] - 1)
```

Feel free to ignore.

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 7]
```diff
@@ -659,19 +655,35 @@ def _apply_embedding_bias(
...
+                                          1)
+            for i in range(len(scheduled_requests.generation_requests)):
+                gen_logits_indices.append(total_context_logits + i)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:677]:
```suggestion
            gen_logits_indices.extend(range(total_context_logits, total_context_logits + len(scheduled_requests.generation_requests)))
```

Feel free to ignore.

File: tensorrt_llm/_torch/pyexecutor/handle_logits.py (hunk 3 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/handle_logits.py:18]: You're a wizard, Harry.

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 6 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:674]: For small lists, list is a more performant type, let's keep a list.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 5 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py:880]: I'll do that in a follow-up PR, so as not to block dev by others.
